### PR TITLE
[Metadata] Extensible metadata component and bundle for Symfony2

### PR DESCRIPTION
**Pull request by pamil**
## Metadata

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | #2307, #2358
| License       | MIT
| Doc PR       | Sylius/Sylius-Docs#322

This PR is aiming to provide an extensible metadata component for PHP and a bundle to integrate it with Symfony2.

For integration with Sylius see #3134.

### Possible improvements and extensions
  - Caching to speed it up
  - More models (Open Graph, Schema.org)
  - More advanced rendering abilities
  - Dynamic expressions (eg. "Products" scope will define`"%PRODUCT_NAME% - %PRODUCT_PRICE%` as title and it will be used for its children, unless they defined their own title)
  - Ability to change page URL (eg. product's url)

### TODO
 - [ ] Documentation

### CODE REVIEW

File: src/Sylius/Bundle/MetadataBundle/DynamicForm/DynamicFormsChoicesMapInterface.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+     * @return string[]
+     */
+    public function getFormsNamesByGroup($group);
+
+    /**
+     * @param string $group
+     * @param string $dataClass
+     *
+     * @return string|null
+     *
```
Comment (peteward) [inline src/Sylius/Bundle/MetadataBundle/DynamicForm/DynamicFormsChoicesMapInterface.php]: blank line!

File: src/Sylius/Bundle/MetadataBundle/Form/Type/Custom/PageMetadataType.php [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+     *
+     * @var DynamicFormBuilderInterface $dynamicFormBuilder
+     */
+    public function __construct($dataClass, array $validationGroups = [], DynamicFormBuilderInterface $dynamicFormBuilder)
+    {
+        parent::__construct($dataClass, $validationGroups);
+
+        $this->dynamicFormBuilder = $dynamicFormBuilder;
+    }
+
```
Comment (peteward) [inline src/Sylius/Bundle/MetadataBundle/Form/Type/Custom/PageMetadataType.php]: blank line

File: src/Sylius/Bundle/MetadataBundle/Form/Type/Twitter/TwitterCardType.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+class TwitterCardType extends AbstractType
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver
+            ->setDefaults([
+            'group' => null,
```
Comment (peteward) [inline src/Sylius/Bundle/MetadataBundle/Form/Type/Twitter/TwitterCardType.php]: indent

File: src/Sylius/Component/Metadata/Model/RootMetadata.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Metadata\Model;
+
```
Comment (peteward) [inline src/Sylius/Component/Metadata/Model/RootMetadata.php]: blank line

File: src/Sylius/Component/Metadata/Provider/MetadataProvider.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        $this->metadataHierarchyProvider = $metadataHierarchyProvider;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getMetadataBySubject(MetadataSubjectInterface $metadataSubject)
+    {
+        $identifiers = $this->getHierarchyByMetadataSubject($metadataSubject);
+
```
Comment (peteward) [inline src/Sylius/Component/Metadata/Provider/MetadataProvider.php]: blank line

File: src/Sylius/Component/Metadata/Provider/MetadataProviderInterface.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
```
Comment (peteward) [inline src/Sylius/Component/Metadata/Provider/MetadataProviderInterface.php]: blank line

**Comment by peteward** [comment]
HI @pamil,

The code looks great to me so far, I'm keen to see how you will extend for the other models and of course the dynamic expressions that you mentioned.

I've tried to get this working locally which has been a challenge so far. The permissions don't appear to be setup for Administrator to access any of the Content section or it's children. The permissions page is also broken relating to CSRF protection so I added the permissions manually to the DB.

You'll want a Metadata menu option in the content section, but I accessed it by URL.
No translations yet so basic labels but I guess that will come later.

I can't add a Metadata now - I get the following error:

PropertyAccessor requires a graph of objects or arrays to operate on, but it found type "NULL" while trying to traverse path "twitter" at property "twitter"

Any ideas?

**Comment by peteward** [comment]
sorry, made this comment on the wrong PR... That's the problem with 3 PRs for the same thing! :)
I have checked out your pamil:seo-sylius-integration branch

**Comment by pamil** [comment]
Hello @peteward,

sorry from delay, just got back from my holidays :) As for permissions, I got the same issue, metadata menu didn't appear on page, it was solved by running `app/console sylius:rbac:initialize`.

I don't know anything about the permissions page, is it related to metadata PRs or independent Sylius bug? If so, I suggest to create a new issue so we won't forget that later :)

In fact, I was thinking about buttons "Edit metadata" on product backend pages, that redirects you to correct metadata edit page. The current approach is meant tosimplify manual testing during development phase, translations will come later too.

I'll investigate your error, this PR is still in progress and there's a lot of work ahead :)

As for your comment in #3127:

At the moment I'm trying to provide *working* solution that offers inheritance for static content and basic rendering. I want to avoid huuge PRs that lags on GitHub and nobody wants to review them :) As soon as the basic functionality will be reached, I'm going to open another PRs with new features like dynamic expressions, caching or some improvements.

I know dynamic expressions and their inheritance is a killer feature for you so it would be done as soon as possible. Products will inherit archetype data for now, as there are many taxons and there is no easy way to implement multiple inheritance, each metadata can have only one parent.

IMHO product variant inheritance should look like that:
 - Default store page metadata
 - Products metadata (applied to every product)
 - Product archetype metadata (if exists)
 - Product metadata
 - Product variant metadata
With that structure, you have many extensions points available and there's no need to do another conditional statements in the product metadata body.

I just came into another issue, that metadata can be used in many contexts. For example, there can be taxon page and product page (with main taxon [not now, may be introduced in the future]) and both of them uses Taxon metadata, but one as for product, one as for taxon. It needs to be solved until first release.

Your proposition about using raw Twig in first iteration is an good idea, I will go for it.

Ouch, that was quite a massive comment, my longest on GitHub :) Hope it'll clarify some doubts and issues.

#3127 is closed, for now please comment only on #3134, it contains all metadata-related stuff I've written.

**Comment by pamil** [comment]
Closing in favour of #3350.

**Pull request closed without merging**